File: src/compiler/scala/tools/nsc/matching/StateSetComparator.scala
/* NSC -- new Scala compiler
 * Copyright 2005-2006 LAMP/EPFL
 * @author Burak Emir
 */
// $Id$

package scala.tools.nsc.matching

import java.util.{Comparator, TreeSet}

class StateSetComparator extends Comparator {

  // use lexicographic order
  def compare(o1: Any, o2: Any): Int = {
    val it1 = o1.asInstanceOf[TreeSet].iterator()
    val it2 = o2.asInstanceOf[TreeSet].iterator()
    while (it1.hasNext()) {
      while (it2.hasNext()) {
        if (!it1.hasNext())
          return -1;

        val i1 = it1.next().asInstanceOf[Integer].intValue()
        val i2 = it2.next().asInstanceOf[Integer].intValue()
        if (i1 < i2)
          return -1;
        else if (i1 > i2)
          return 1;
      }
      if (it1.hasNext())
        return 1;
    }
    if (it2.hasNext()) -1 else 0
  }
}

File: sources/scalac/transformer/matching/StateSetComparator.java
package scalac.transformer.matching ;

import java.util.* ;

class StateSetComparator implements Comparator {
            // use lexicographic order
            public int compare( Object o1, Object o2 ) {
                  /*
                  System.out.print("lexi" );
                  System.out.print( o1 +" ");
                  System.out.println( o2 );
                  */
                  Iterator it1 = ((TreeSet) o1).iterator();
                  Iterator it2 = ((TreeSet) o2).iterator();
                  while( it1.hasNext() ) {
                        while( it2.hasNext() ) {
                              if( !it1.hasNext() )
                                    return -1;

                              int i1 = ((Integer) it1.next()).intValue();
                              int i2 = ((Integer) it2.next()).intValue();
                              if( i1 < i2 )
                                    return -1;
                              else if ( i1 > i2 )
                                    return 1;
                        }
                        if( it1.hasNext() )
                              return 1;
                  }
                  if( it2.hasNext() )
                        return -1;
                  return 0;
            }
      }
